### PR TITLE
US001-view-job-roles

### PR DESCRIPTION
**Pull request by ghost**
API endpoint to view job roles

### CODE REVIEW

**Review by leeservice — APPROVED** [review]
I've looked through this, made one change to the DBConnection and am happy - @ccurran01  can you please review

File: src/main/java/org/kainos/ea/validator/JobSpecValidator.java [hunk 1]
```diff
@@ -1,6 +1,4 @@
 package org.kainos.ea.validator;
 
 public class JobSpecValidator {
```
Comment (ccurran01) [inline src/main/java/org/kainos/ea/validator/JobSpecValidator.java:3]: Do we need this to view a list of job roles?

File: src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class JobRoleServiceTest {
```
Comment (ccurran01) [inline src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java:24]: I think we are missing a scenario here for when the database connector throws an exception?

File: src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import org.kainos.ea.trueConfiguration;
+
+import java.util.List;
+
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class JobRoleControllerTest {
+    static final DropwizardAppExtension<trueConfiguration> APP = new DropwizardAppExtension<>(trueApplication.class, null, new ResourceConfigurationSourceProvider());
+
+    @Test
+    void getJobRoles_shouldReturnListOfJobRoles() {
```
Comment (ccurran01) [inline src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java:20]: Are we able to assert the status code of the response as well? Should be a 200? https://http.cat/200

**Review by ccurran01 — COMMENTED** [review]
Looking good so far 👍, there's some unused code we can tidy up and some extra test coverage to add

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+@Api("Commit Connoisseurs API")
+@Path("/api")
+public class JobRoleController {
+    private static JobRoleService jobRoleService;
+    private static JobRoleValidator jobRoleValidator;
+
+    public JobRoleController() {
+        DatabaseConnector databaseConnector = new DatabaseConnector();
+        jobRoleService = new JobRoleService(new JobRoleDao(), databaseConnector);
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/controller/JobRoleController.java:24]: cleaner and implemented

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        jobRoleValidator = new JobRoleValidator();
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/controller/JobRoleController.java:25]: No, that's superfluous right now, good spot, removed and updated

File: empty.txt
[the remark this answers is not in the document]
Comment (leeservice) [reply empty.txt:1]: Absolutely not, deleted

File: src/main/java/org/kainos/ea/model/Band.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
+package org.kainos.ea.model;
+
+public class Band {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/model/Band.java:3]: Not fully, but it's required for user story 3 or 4 for future sprints and was implemented into the front end already.

File: src/main/java/org/kainos/ea/model/JobRoleRequest.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package org.kainos.ea.model;
+
+public class JobRoleRequest {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/model/JobRoleRequest.java:3]: It was to be created for getting view band level and view capability - not relevant to this userstory :(

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import org.kainos.ea.exceptions.FailedToGetJobRoleException;
+import org.kainos.ea.model.JobRoleRequest;
+import org.kainos.ea.utility.DatabaseConnector;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+
+public class JobRoleService {
+    public JobRoleDao jobRoleDao = new JobRoleDao();
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/service/JobRoleService.java:14]: fair!

File: src/main/java/org/kainos/ea/service/JobSpecService.java [hunk 9]
```diff
@@ -1,5 +1,4 @@
 package org.kainos.ea.service;
 
 public class JobSpecService {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/service/JobSpecService.java:3]: Not relevant to this userstory

File: src/main/java/org/kainos/ea/controller/JobSpecController.java [hunk 10]
```diff
@@ -3,10 +3,9 @@
 import io.swagger.annotations.Api;
 import javax.ws.rs.Path;
 import org.kainos.ea.service.JobSpecService;
-import org.kainos.ea.utility.DatabaseConnector;
 import org.kainos.ea.validator.JobSpecValidator;
 
-@Api("Kainos Job Application")
+@Api("Commit Connoisseurs API")
 @Path("/job-specification")
 public class JobSpecController {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/controller/JobSpecController.java:10]: Not relevant to this userstory

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package org.kainos.ea.validator;
+
+import org.kainos.ea.model.JobRole;
+
+public class JobRoleValidator {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/validator/JobRoleValidator.java:5]: No :(

File: src/main/java/org/kainos/ea/validator/JobSpecValidator.java (hunk 1 above)
Comment (leeservice) [reply src/main/java/org/kainos/ea/validator/JobSpecValidator.java:3]: No, it's for other user stories

Comment (leeservice) [reply src/main/java/org/kainos/ea/validator/JobSpecValidator.java:3]: I've deleted this as we're discussing changes

File: src/main/java/org/kainos/ea/dao/JobRoleDao.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        ResultSet rs = st.executeQuery("SELECT Job_Roles.id AS 'ID', Job_Roles.`name` AS 'Name', description AS 'Specification Description'," +
+                " level_of_band AS 'Band Level', Capability.`name` AS 'Capability'" +
+                " FROM Job_Roles" +
+                " INNER JOIN Banding ON Job_Roles.band_id = Banding.id" +
+                " INNER JOIN Capability ON Job_Roles.capability_id = Capability.id;");
+
+        List<JobRoleRequest> jobRoleList = new ArrayList<>();
+
+        while(rs.next()) {
+            JobRoleRequest jobRoleRequest = new JobRoleRequest(
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/dao/JobRoleDao.java:26]: I've updated this and removed all jobrolerequest files - it's now all attuned to jobrole, as this is relevant to user story

File: src/main/java/org/kainos/ea/model/JobRoleRequest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/model/JobRoleRequest.java:3]: I've removed it

File: src/main/java/org/kainos/ea/service/JobSpecService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/service/JobSpecService.java:3]: I've removed it

File: src/main/java/org/kainos/ea/controller/JobSpecController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/controller/JobSpecController.java:10]: I've removed it

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/validator/JobRoleValidator.java:5]: No, but it is good for rules

File: src/main/java/org/kainos/ea/model/Band.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (leeservice) [reply src/main/java/org/kainos/ea/model/Band.java:3]: I've removed this to focus on this userstory

File: src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java (hunk 2 above)
Comment (leeservice) [reply src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java:24]: We've added this now.

File: src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.kainos.ea.model.JobRoleRequest;
+import org.kainos.ea.trueApplication;
+import org.kainos.ea.trueConfiguration;
+
+import java.util.List;
+
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class JobRoleControllerTest {
```
[the remark this answers is not in the document]
Comment (leeservice) [reply src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java:16]: Implemented

File: src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java (hunk 3 above)
Comment (leeservice) [reply src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java:20]: Done

File: src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    @Test
+    void getJobRoles_shouldThrowFailedToGetJobRoleException_whenDaoThrowsSQLException() throws SQLException, DatabaseConnectionException {
+        when(databaseConnector.getConnection()).thenReturn(conn);
+        when(jobRoleDao.getAllJobRoles(conn)).thenThrow(SQLException.class);
+
+        assertThrows(FailedToGetJobRoleException.class, () -> jobRoleService.getAllJobRoles());
+    }
+
+    @Test
+    void getJobRoles_shouldThrowFailedToGetJobRoleException_whenDaoThrowsDatabaseConnectionException() throws SQLException, DatabaseConnectionException {
```
Comment (ccurran01) [inline src/test/java/org/kainos/ea/serviceTests/JobRoleServiceTest.java:40]: I don't think the DAO throws a database connection exception. Should this be the database connector instead?

File: src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Test
+    void getJobRoles_shouldReturnListOfJobRoles() {
+        List<JobRole> response =
+                APP.client()
+                        .target("http://localhost:8080/api/job-roles")
+                        .request()
+                        .get(List.class);
+        Assertions.assertTrue(response.size() > 0);
+    }
```
[the remark this answers is not in the document]
Comment (ccurran01) [reply src/test/java/org/kainos/ea/integrationTests/JobRoleControllerTest.java]: This is showing as outdated but I think its still valid

**Review by ccurran01 — APPROVED** [review]

**Pull request merged**